### PR TITLE
adding time_spec to mg_http_serve_opts for choosing how to display last modified date

### PR DESCRIPTION
**Pull request by behnejad**
adding time_spec to mg_http_serve_opts for choosing how to display last modified date.

### CODE REVIEW

**Comment by scaprile** [comment]
Since they're mutually exclusive, I vote for having a single `MG_HTTP_DIRLIST_TIME` with possible values
```
enum {
    MG_TIMESPEC_OFF = 0 ,
    MG_TIMESPEC_LOCAL,
    MG_TIMESPEC_UTC
}
```
That can also be used in other places if and when necessary

**Comment by cpq** [comment]
why not strftime format string? When NULL, we'll use a default.

```c
  const char *time_fmt;
```

File: src/http.c [hunk 1]
```diff
@@ -562,23 +562,20 @@ static void printdirentry(const char *name, void *userdata) {
...
-      char time_str[30];
+    } else if (d->opts->time_spec == MG_TIMESPEC_UTC) {
+      char time_str[MG_HTTP_TIME_FMT_SIZE];
       struct tm *time_info = gmtime(&t);
-      strftime(time_str, sizeof time_str, "%Y/%m/%d %H:%M:%S", time_info);
+      strftime(time_str, sizeof time_str, d->opts->time_fmt ? d->opts->time_fmt : "%Y/%m/%d %H:%M:%S", time_info);
       mg_snprintf(mod, sizeof(mod), "%s", time_str);
+    } else {
+      mg_snprintf(mod, sizeof(mod), "%ld", (unsigned long) t);
     }
```
Comment (scaprile) [inline src/http.c:577]: Perhaps
```
if (d->opts->time_spec == MG_TIMESPEC_OFF) {
  mg_snprintf(mod, sizeof(mod), "%ld", (unsigned long) t);
} else {
  char time_str[MG_HTTP_TIME_FMT_SIZE];
  struct tm *time_info = (d->opts->time_spec == MG_TIMESPEC_LOCAL) ? localtime(&t) : gmtime(&t);
  strftime(time_str, sizeof(time_str), d->opts->time_fmt ? d->opts->time_fmt : "%Y/%m/%d %H:%M:%S", time_info);
  mg_snprintf(mod, sizeof(mod), "%s", time_str);
}
```
?

**Review by scaprile — COMMENTED** [review]
Looks nice

**Comment by cpq** [comment]
Please get rid of everything that is not necessary. MG_TIMESPEC_ENUM should go.  MG_HTTP_TIME_FMT_SIZE should go.
Make a buffer 100 bytes long, do not introduce yet another tunable - we have enough of them. `time_fmt` is the only thing that should be introduced by this PR into the API, nothing else.

**Comment by scaprile** [comment]
How do you suggest to choose between seconds, formatted local, and formatted UTC without the enum ?

**Comment by behnejad** [comment]
I think displaying time in seconds is not necessary, since we want it to be human readable so MG_TIMESPEC_OFF is not needed.
gmtime and localtime can be chosen at compile time or hard coded.
Time format can also be hard coded or chosen at compile time or defined dynamically by developer with default format (personally prefer).
It is possible to let user set time function with registering a callback, if it is not set, localtime will be default callback.

**Comment by cpq** [comment]
I like the idea of the compile-time format option!
A dynamic format, IMO, is an overkill.
Then, we could have a build option `MG_DIR_LIST_TIME_FMT` set to "%Y/%m/%d %H:%M:%S" by default?

Or maybe we just don't do any of that, and use time format that is traditionally used by other servers? @behnejad what exactly made you do this PR? What format you prefer to see?

**Comment by behnejad** [comment]
I like to use "%Y/%m/%d %H:%M:%S" format with localtime and by seeing other webservers, they have hard coded format.
Before this PR, I opened another one to add time format, by the time i though it is good to have an option to choose which time (local / gm) to use. But now i think it is overkill like you said.
How about having a macro for time format and use localtime only without any options?

**Comment by scaprile** [comment]
> I think displaying time in seconds is not necessary, since we want it to be human readable so MG_TIMESPEC_OFF is not needed.

We actually have systems that don't have the notion of that, only uptime is available. Why would those systems want to show a directory is another question, but if possible I would like to keep the unformatted plain uptime option, as formatting that may send us back to the loved 70's and I'm too old to rock'n'roll.

**Pull request closed without merging**